sdk/go/common/util/gitutil/git.go

```diff
@@ -857,6 +862,52 @@ func GitListBranchesAndTags(url string) ([]plumbing.ReferenceName, error) {
 	return results, nil
 }
 
+// GetLatestTagOrHash returns the latest tag or hash in the repository.
+// To do this, we list all the tags in the repository and try to pars them as semver.  If we can't
+// find any valid tags that are semver, we'll use the hash of the default branch.  If we can't find
+// any tags or default branches we return an error.
+func GetLatestTagOrHash(ctx context.Context, url string) (*semver.Version, error) {
+	refs, err := gitListRefs(ctx, url)
+	if err != nil {
+		return nil, err
+	}
+
+	var versions []semver.Version
+	namedHashes := make(map[plumbing.ReferenceName]plumbing.Hash)
+	var headRef plumbing.ReferenceName
+	for _, ref := range refs {
+		name := ref.Name()
+		if name.IsTag() {
+			tag := name.Short()
+			v, err := semver.ParseTolerant(tag)
+			if err == nil {
+				versions = append(versions, v)
+			}
+		} else if name.IsBranch() {
+			namedHashes[name] = ref.Hash()
+		} else if name == plumbing.HEAD {
+			headRef = ref.Target()
+		}
+	}
+	if len(versions) == 0 {
+		if hash, ok := namedHashes[headRef]; ok {
```

Comment:
Could it be that we couldn't find a HEAD? Should we assert on this explicitly?

Reply:
I don't think that should be the case, and in that case we would just end up in the error case below.  I dunno if we should should explicitly specify that, I think the only reason that would happen is if the repo was completely empty, or if the repo is in a weird state.  Would we give a different error message than `could not determine version in repo` for those?

Reply:
Yeah probably that's fine as-is